local-cluster/tests/local_cluster.rs

```diff
@@ -4480,6 +4480,216 @@ fn test_slot_hash_expiry() {
     );
 }
 
+// This tests vote refreshes on votes that are too old to be in slot hash history. Such votes
+// will never land, so we should instead send a new vote for the tip of the heaviest fork
+// descending from the old vote.
+// It creates validators A, B, and C, (C is only needed for stake distribution), then do this:
+// 1. kill B and C, let A run for a while to create its own fork A
+// 2. kill A, copy ledger of A to B so B has everything until the common ancestor, restart B and
+//    let it run long enough to create a new fork B
+// 3. bring back A and B only, A now has 37% of the stake so its fork is heavier, B has only 35%
+//    so it wants to switch to fork A, but can't do it because of A's stake under switch proof
+//    threshold of 38%. Also, B has voted on fork B, and last_voted < last_landed_vote (we do this
+//    by stopping B after its 4-block leader slot), and last vote is old enough so it's outside
+//    slothash
+// 4. we check that B will be able to vote again and the new vote is further down the B fork.
+#[test]
+#[serial]
+fn test_vote_refresh_outside_slothash() {
+    solana_logger::setup_with_default(RUST_LOG_FILTER);
+    solana_sdk::slot_hashes::set_entries_for_tests_only(8);
```

Comment:
Using a constant here `const MAX_SLOT_HASHES = 8` would be helpful for reusing `MAX_SLOT_HASHES` later for integrity checks

Reply:
Done.